zipline/pipeline/data/dataset.py

```diff
@@ -160,6 +166,28 @@ def _static_identity(cls, dataset, name, doc, metadata, *args, **kwargs):
             frozenset(sorted(metadata.items(), key=first)),
         )
 
+    def specialize(self, domain):
+        """Specialize ``self`` to a concrete domain.
+        """
+        if self.domain == domain:
```

Comment:
Most of the time when we're specializing stuff, our logic is essentially "take a mix of generic and already-specialized terms, and make sure they're all specialized to an actual domain", so having this work allows us to just write:

`[term.specialize(domain) for term in terms]`

instead of something more verbose like:

`[term.specialize(domain) if term.domain is not domain else term for term in terms]`